agg-ast/schema_derivation/src/schema_derivation.rs

```diff
@@ -1747,6 +1746,70 @@ impl DeriveSchema for UntaggedOperator {
+            let mut type_set: BTreeSet<Schema> = set!();
+            let mut arg_schemas: BTreeSet<Schema> = set!();
+            // here we derive schema for all args and filter for numeric types only.
+            // we also use this time to capture if any args MAY be nullable, and to
+            // return null if any arg MUST be null or missing
+            for arg in args.iter() {
+                let arg_schema = arg.derive_schema(state)?;
+                match arg_schema.satisfies(&NULLISH.clone()) {
+                    Satisfaction::Must => return Ok(Schema::Atomic(Atomic::Null)),
+                    Satisfaction::May => {
+                        type_set.insert(Schema::Atomic(Atomic::Null));
+                    }
+                    Satisfaction::Not => {}
+                }
+                arg_schemas.insert(arg_schema.intersection(&NUMERIC.clone()));
+            }
+            // we can set the lower bound of numeric args as the largest type that is the minimum
+            // of one of the arguments. For example, if the lowest ordered type an argument can take
+            // on is a Double, the set that as the lower bound, as the operator will never return int or long.
+            let type_bound = arg_schemas
+                .iter()
+                .map(|schema| match schema {
+                    Schema::Atomic(_) => schema,
+                    Schema::AnyOf(ao) => ao.iter().min().unwrap(),
+                    _ => &Schema::Unsat,
+                })
+                .max()
+                .unwrap()
+                .to_owned();
+            type_set.insert(type_bound.clone());
+            // we then add any possible types that are > the lower bound to the set of types
+            // our operator can return, and return the anyof of all the possible types.
+            for arg_schema in arg_schemas {
+                match arg_schema {
+                    Schema::Atomic(_) => {
+                        if arg_schema > type_bound {
+                            type_set.insert(arg_schema);
+                        }
+                    }
+                    Schema::AnyOf(ao) => {
+                        ao.into_iter().for_each(|schema| {
+                            if schema > type_bound {
+                                type_set.insert(schema);
+                            }
+                        });
+                    }
+                    _ => {}
+                }
+            }
+            let mut schema = Schema::simplify(&Schema::AnyOf(type_set));
+            // the operators calling this, like multiply and add, will return a long if the
+            // inputs are integers and the value is sufficiently large, so we add that type
+            if schema.satisfies(&Schema::Atomic(Atomic::Integer)) != Satisfaction::Not
+                && schema != Schema::Unsat
+            {
+                schema = schema.union(&Schema::Atomic(Atomic::Long));
+            }
+            Ok(schema)
```

Comment:
1. Do we need to simplify again here? If `schema` was `AnyOf([Int, Long])` already, will the union result in `AnyOf([Int, Long, Long])`?
2. We need to handle overflow for all types (int overflow -> long, long overflow -> double, double/decimal overflow -> also double (I'm pretty sure +/- inf is always a double although we should check that). ([link](https://www.mongodb.com/docs/manual/reference/operator/aggregation/add/#behavior))

Reply:
1. union of anyof and atomic with insert the atomic directly into the underlying btree set. so we're good there :) 
2. Decimal does have an inf type, which came up a lot in the fuzz tests. I'll add long overflow to double and add a similar test case to the one you described below

Reply:
1. Oh right right right, embarrassing 😅 of course we use a set not a vec for that
2. Ok thank you